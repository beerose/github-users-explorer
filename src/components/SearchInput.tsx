/** @jsx jsx */
import React, { ComponentProps, useRef } from "react";
import { css, jsx } from "@emotion/core";

import { styled, useTheme } from "../theme";
import { transparentize } from "polished";
import { useFocus } from "../utils";

import { SearchIcon, XIcon } from "./icons";
import { Button } from "./Button";

type ClearInputButtonProps = ComponentProps<"button">;
const ClearInputButton: React.FC<ClearInputButtonProps> = props => {
  const { colors } = useTheme();

  return (
    <Button
      title="Clear input"
      css={css`
        display: flex;

        color: ${colors.text};
        opacity: 0.5;

        outline: none;

        :hover,
        :focus {
          opacity: 1;
        }
      `}
      {...props}
    >
      <XIcon width="1em" height="1em" />
    </Button>
  );
};
type StyledLabelProps = { inputFocused: boolean };
const Label = styled.label<StyledLabelProps>`
  display: flex;

  font-size: 2em;
  padding: 0 1em;

  background: white;
  border-radius: 1.6em;

  ${({ theme, inputFocused }) =>
    inputFocused &&
    css`
      box-shadow: 0 0 15px 1px ${transparentize(0.95, theme.colors.text)},
        0 0 1px 1px ${transparentize(0.88, theme.colors.text)};
    `};

  @media (max-width: ${({ theme }) => theme.breakpoints.mobile}) {
    font-size: 1.2em;
  }
`;

const Input = styled.input`
  border: none;
  outline: none;

  overflow: hidden;
  text-overflow: ellipsis;

  background: none;

  flex: 1 1 0;

  font-size: inherit;
  line-height: 1.7;

  ::placeholder {
    opacity: 0.5;
  }
`;

interface SearchInputProps extends Omit<ComponentProps<"input">, "onReset"> {
  onReset: () => void;
}
export const SearchInput: React.FC<SearchInputProps> = ({
  className,
  onReset,
  ...inputProps
}) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const inputFocused = useFocus(inputRef);
  const { colors } = useTheme();

  return (
    <Label inputFocused={inputFocused} className={className}>
      <SearchIcon
        width="1em"
        height="100%"
        stroke={transparentize(0.5, colors.text)}
        css={css`
          margin-right: 0.5em;
        `}
      />
      <Input ref={inputRef} {...inputProps} />
      <ClearInputButton onClick={onReset} />
    </Label>
  );
};
