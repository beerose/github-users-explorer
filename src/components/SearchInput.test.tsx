import React from "react";
import "@testing-library/jest-dom/extend-expect";

import { render } from "../../test/testUtils";
import { SearchInput } from "./SearchInput";

describe("SearchInput", () => {
  const username = "jondoe";
  const noop = () => null;

  it("displays username in the input field", () => {
    const { getByDisplayValue } = render(
      <SearchInput value={username} onReset={noop} onChange={noop} />
    );

    expect(getByDisplayValue(username)).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <SearchInput value={username} onReset={noop} onChange={noop} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
