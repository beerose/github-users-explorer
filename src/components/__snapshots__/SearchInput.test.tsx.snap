// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput matches snapshot 1`] = `
<DocumentFragment>
  <label
    class="css-kqqiw6"
  >
    <svg
      class="css-jyxcdo"
      fill="none"
      height="100%"
      stroke="rgba(10,9,12,0.5)"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
    >
      <mask
        id="SearchIcon_clipPath"
      >
        <rect
          fill="white"
          height="100%"
          width="100%"
          x="0"
          y="0"
        />
        <circle
          cx="11"
          cy="11"
          fill="black"
          r="8"
          stroke="black"
        />
      </mask>
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <path
        d="M21 21L16.65 16.65"
        mask="url(#SearchIcon_clipPath)"
      />
    </svg>
    <input
      class="css-xi6tab"
      value="jondoe"
    />
    <button
      class="css-1aktd5y-ClearInputButton"
      title="Clear input"
    >
      <svg
        class="feather feather-x"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M18 6L6 18 M6 6L18 18"
        />
      </svg>
    </button>
  </label>
</DocumentFragment>
`;
